File: python/examples/parse_declarations.py
#!/usr/bin/env python3
# Copyright (c) 2024-2026 Elias Bachaalany
# SPDX-License-Identifier: MPL-2.0
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at https://mozilla.org/MPL/2.0/.
#
# parse_declarations: Import C type declarations, verify, clean up.
#
# Usage: python parse_declarations.py [host_url]
#
# Demonstrates bulk C type import via Ghidra's CParser.

import sys

import libghidra as ghidra


def main() -> None:
    url = sys.argv[1] if len(sys.argv) >= 2 else "http://127.0.0.1:18080"
    client = ghidra.connect(url)

    # 1. Verify connection
    try:
        status = client.get_status()
    except ghidra.GhidraError as e:
        print(f"Cannot reach host at {url}: {e}", file=sys.stderr)
        sys.exit(1)

    print(f"Connected: {status.service_name} v{status.service_version}")

    # 2. Parse a block of C declarations
    decls = """\
typedef enum ExampleOpcode {
    OP_NONE = 0,
    OP_INIT = 1,
    OP_PROCESS = 2,
    OP_SHUTDOWN = 3
} ExampleOpcode;

typedef struct ExampleHeader {
    int magic;
    int version;
    int flags;
} ExampleHeader;

typedef struct ExamplePacket {
    ExampleHeader header;
    ExampleOpcode opcode;
    int payload_size;
} ExamplePacket;
"""

    print("\n--- Parsing C declarations ---")
    try:
        resp = client.parse_declarations(decls)
        print(f"Types created: {resp.types_created}")
        for name in resp.type_names:
            print(f"  + {name}")
        if resp.errors:
            print("Errors:")
            for err in resp.errors:
                print(f"  ! {err}")
    except ghidra.GhidraError as e:
        print(f"ParseDeclarations failed: {e}", file=sys.stderr)
        sys.exit(1)

    # 3. Verify the types exist in the type system
    print("\n--- Verifying types ---")
    check_names = ["/ExampleOpcode", "/ExampleHeader", "/ExamplePacket"]
    for name in check_names:
        try:
            resp = client.get_type(name)
            if resp.type:
                t = resp.type
                print(f"  {name}: kind={t.kind} length={t.length}")
            else:
                print(f"  {name}: NOT FOUND")
        except ghidra.GhidraError:
            print(f"  {name}: NOT FOUND")

    # 4. Clean up: delete the types we created
    print("\n--- Cleanup ---")
    for name in reversed(check_names):
        try:
            client.delete_type(name)
        except ghidra.GhidraError:
            pass
    print("  Deleted all example types")

    print("\nDone.")


if __name__ == "__main__":
    main()

File: cpp/examples/parse_declarations.cpp
// Copyright (c) 2024-2026 Elias Bachaalany
// SPDX-License-Identifier: MPL-2.0
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// parse_declarations: Import C type declarations, verify they exist, clean up.
//
// Usage: parse_declarations [host_url]

#include <cstdio>
#include <cstdlib>
#include <string>

#include "libghidra/ghidra.hpp"

int main(int argc, char* argv[]) {
  const std::string host_url =
      (argc >= 2) ? argv[1] : "http://127.0.0.1:18080";

  auto client = ghidra::connect(host_url);

  // 1. Check host health
  auto status = client->GetStatus();
  if (!status.ok()) {
    fprintf(stderr, "Cannot reach host at %s: %s\n", host_url.c_str(),
            status.status.message.c_str());
    return 1;
  }
  printf("Connected: %s v%s\n",
         status.value->service_name.c_str(),
         status.value->service_version.c_str());

  // 2. Parse a block of C declarations
  const char* decls = R"(
    typedef enum ExampleOpcode {
      OP_NONE = 0,
      OP_INIT = 1,
      OP_PROCESS = 2,
      OP_SHUTDOWN = 3
    } ExampleOpcode;

    typedef struct ExampleHeader {
      int magic;
      int version;
      int flags;
    } ExampleHeader;

    typedef struct ExamplePacket {
      ExampleHeader header;
      ExampleOpcode opcode;
      int payload_size;
    } ExamplePacket;
  )";

  printf("\n--- Parsing C declarations ---\n");
  auto result = client->ParseDeclarations(decls);
  if (!result.ok()) {
    fprintf(stderr, "ParseDeclarations failed: %s\n",
            result.status.message.c_str());
    return 1;
  }
  printf("Types created: %d\n", result.value->types_created);
  for (const auto& name : result.value->type_names) {
    printf("  + %s\n", name.c_str());
  }
  if (!result.value->errors.empty()) {
    printf("Errors:\n");
    for (const auto& err : result.value->errors) {
      printf("  ! %s\n", err.c_str());
    }
  }

  // 3. Verify the types exist in the type system
  printf("\n--- Verifying types ---\n");
  const char* check_names[] = {
      "/ExampleOpcode", "/ExampleHeader", "/ExamplePacket"};
  for (const auto& name : check_names) {
    auto t = client->GetType(name);
    if (t.ok() && t.value->type) {
      printf("  %s: kind=%s length=%llu\n", name,
             t.value->type->kind.c_str(),
             (unsigned long long)t.value->type->length);
    } else {
      printf("  %s: NOT FOUND\n", name);
    }
  }

  // 4. Clean up: delete the types we created
  printf("\n--- Cleanup ---\n");
  client->DeleteType("/ExamplePacket");
  client->DeleteType("/ExampleHeader");
  client->DeleteType("/ExampleOpcode");
  printf("  Deleted all example types\n");

  return 0;
}
